cost-model/src/cost_tracker.rs

```diff
@@ -160,16 +225,54 @@ impl CostTracker {
         self.in_flight_transaction_count -= in_flight_transaction_count
     }
 
-    pub fn try_add(
+    pub fn try_stage(
         &mut self,
         tx_cost: &TransactionCost<impl TransactionWithMeta>,
+        tx_index: usize,
     ) -> Result<UpdatedCosts, CostTrackerError> {
-        self.would_fit(tx_cost)?;
-        let updated_costliest_account_cost = self.add_transaction_cost(tx_cost);
-        Ok(UpdatedCosts {
+        // this assertion will only fail if the validator code itself has a bug that
+        // causes transaction costs to be processed out of order.
+        assert_eq!(tx_index as u64, self.transaction_count());
+
+        let (track_index, start_cus) = self.try_fit(tx_cost)?;
+
+        let updated_costliest_account_cost =
+            self.add_transaction_cost(tx_cost, tx_index, track_index, start_cus);
+
+        let updated_costs = UpdatedCosts {
             updated_block_cost: self.block_cost,
             updated_costliest_account_cost,
-        })
+            // todo: add serial execution cost
+        };
+
+        self.pending_cost_updates.push(updated_costs.clone());
+        Ok(updated_costs)
+    }
+
+    pub fn try_stage_next(
+        &mut self,
+        tx_cost: &TransactionCost<impl TransactionWithMeta>,
+    ) -> Result<UpdatedCosts, CostTrackerError> {
+        self.try_stage(tx_cost, self.transaction_count() as usize)
```

Comment:
`transaction_count()` isn't guaranteed to be linear - `remove_transaction_cost()` reduces it by 1.